File: test/debug/diff_fea_vec.py
import pandas as pd


path_block_stat = "output/blk_res/stat.txt"
with open(path_block_stat, "r") as stat_file:
    stat_line = stat_file.readlines()
    total_table, _ = (int(val) for val in stat_line[0].split())
    
tmp_df = pd.read_csv("output/blk_res/feature_vec0.csv")
schemas = list(tmp_df.columns)
diff_ori_fea_vec = pd.DataFrame(columns=schemas)
diff_pro_fea_vec = pd.DataFrame(columns=schemas)
    
for i in range(total_table):
    print(f"processing feature vector {i} ...")
    path_ori_fea_vec = "output/blk_res/feature_vec" + str(i) + "_py.csv"
    path_pro_fea_vec = "output/blk_res/feature_vec" + str(i) + ".csv"
    
    ori_fea_vec = pd.read_csv(path_ori_fea_vec)
    pro_fea_vec = pd.read_csv(path_pro_fea_vec)
    
    ori_fea_vec.rename(columns={"_id": "id"}, inplace=True)
    
    ori_row_index = list(ori_fea_vec.index)
    pro_row_inedx = list(pro_fea_vec.index)
    length = len(ori_row_index)
    
    for j in range(length):
        ori_ridx = ori_row_index[j]
        pro_ridx = pro_row_inedx[j]
        for attr in schemas:
            if ori_fea_vec.loc[ori_ridx, attr] != pro_fea_vec.loc[pro_ridx, attr] or \
               pd.isna(ori_fea_vec.loc[ori_ridx, attr]) ^ pd.isna(pro_fea_vec.loc[pro_ridx, attr]) == True:
            #    diff_ori_fea_vec[-1] = list(ori_fea_vec.loc[ori_ridx])
            #    diff_pro_fea_vec[-1] = list(pro_fea_vec.loc[pro_ridx])
            #    diff_ori_fea_vec.index = diff_ori_fea_vec.index + 1
            #    diff_pro_fea_vec.index = diff_pro_fea_vec.index + 1
            #    diff_ori_fea_vec.sort_index(inplace=True)
            #    diff_pro_fea_vec.sort_index(inplace=True)
                   
                diff_ori_fea_vec = pd.concat([pd.DataFrame([list(ori_fea_vec.loc[ori_ridx])], columns=diff_ori_fea_vec.columns), diff_ori_fea_vec], ignore_index=True)
                diff_pro_fea_vec = pd.concat([pd.DataFrame([list(pro_fea_vec.loc[pro_ridx])], columns=diff_pro_fea_vec.columns), diff_pro_fea_vec], ignore_index=True)

diff_ori_fea_vec.to_csv("test/debug/diff_ori_fea_vec.csv", index=False)
diff_pro_fea_vec.to_csv("test/debug/diff_pro_fea_vec.csv", index=False)

File: test/debug/diff_fea_vec.cc
#include "common/io.h"
#include <string>


int main()
{
    std::string pathStatFile = "output/blk_res/stat.txt";
    FILE *fp = fopen(pathStatFile.c_str(), "r");

    int totalTable, totalEntity;
    fscanf(fp, "%d %d\n", &totalTable, &totalEntity);
    printf("%d\n", totalTable);

    fclose(fp);

    CSVReader reader;
    auto diffOriFeaVec = Table();
    auto diffProFeaVec = Table();

    for(int i = 0; i < totalTable; i++) {
        std::string pathOriFeaVec = "output/blk_res/feature_vec" + std::to_string(i) + "_py.csv";
        std::string pathProFeaVec = "output/blk_res/feature_vec" + std::to_string(i) + ".csv";

        reader.reading_one_table(pathOriFeaVec, false);
        reader.reading_one_table(pathProFeaVec, false);

        Table oriFeaVec = reader.tables[i * 2];
        Table proFeaVec = reader.tables[i * 2 + 1];

        if(i == 0) {
            diffOriFeaVec.schema = oriFeaVec.schema;
            diffProFeaVec.schema = proFeaVec.schema;
            diffOriFeaVec.inverted_schema = oriFeaVec.inverted_schema;
            diffProFeaVec.inverted_schema = proFeaVec.inverted_schema;
            diffOriFeaVec.cols.resize(oriFeaVec.col_no);
            diffProFeaVec.cols.resize(proFeaVec.col_no);
        }

        oriFeaVec.Profile();
        proFeaVec.Profile();
        int numRow = std::min(oriFeaVec.row_no, proFeaVec.row_no);
        int numColumn = oriFeaVec.col_no;

        for(int row = 0; row < numRow; row++) {
            for(int col = 0; col < numColumn; col++) {
                if(oriFeaVec.rows[row][col].empty() ^ proFeaVec.rows[row][col].empty() || 
                   oriFeaVec.rows[row][col] != proFeaVec.rows[row][col]) {
                    diffOriFeaVec.insertOneRow(oriFeaVec.rows[row]);
                    diffProFeaVec.insertOneRow(proFeaVec.rows[row]);
                    break;
                }
            }
        }
    }

    diffOriFeaVec.Profile();
    diffProFeaVec.Profile();

    MultiWriter::writeOneTable(diffOriFeaVec, "test/debug/ori_fea_vec.csv");
    MultiWriter::writeOneTable(diffProFeaVec, "test/debug/pro_fea_vec.csv");

    return 0;
}